care/facility/events/handler.py

```diff
@@ -0,0 +1,129 @@
+        else:
+            data[field.name] = value
+    return data
+
+
+@shared_task
+def create_consultation_event_task(
+    consultation_id: int,
+    object_model: str,
+    object_id: int,
+    caused_by: int,
+    created_date: datetime,
+    diff: str = None,
+):
+    Model = apps.get_model("facility", object_model)
+
+    print(f"Creating Event for {object_model} {object_id}")
+
+    instance = Model.objects.get(id=object_id)
+
+    diff = json.loads(diff) if diff else None
+    change_type = ChangeType.UPDATED if diff else ChangeType.CREATED
+
+    data = transform(instance, diff)
+    print(data)
+
+    object_fields = set(data.keys())
+
+    with transaction.atomic():
+        batch = []
+        groups = EventType.objects.filter(
+            model=object_model, fields__len__gt=0
+        ).values_list("id", "fields")
+        print(groups)
+        for group_id, group_fields in groups:
+            if set(group_fields) & object_fields:
+                # PatientConsultationEvent.objects.select_for_update().filter(
+                #     consultation_id=consultation_id,
+                #     event_type=group_id,
+                #     is_latest=True,
+                #     created_date__lt=created_date,
+                # ).update(is_latest=False)
+                value = {}
+                for field in group_fields:
+                    try:
+                        value[field] = data[field]
+                    except KeyError:
+                        value[field] = getattr(instance, field, None)
+                batch.append(
+                    PatientConsultationEvent(
+                        consultation_id=consultation_id,
+                        caused_by_id=caused_by,
+                        event_type_id=group_id,
+                        is_latest=True,
+                        created_date=created_date,
+                        object_model=object_model,
+                        object_id=object_id,
+                        value=value,
+                        change_type=change_type,
+                        meta={
+                            "external_id": str(getattr(instance, "external_id", ""))
+                            or None
+                        },
+                    )
+                )
+
+        # print([x.__dict__ for x in batch])
+
+        PatientConsultationEvent.objects.bulk_create(batch)
+        return len(batch)
+
+
+def create_consultation_event(
+    consultation_id: int,
+    objects: list | QuerySet | Model,
+    caused_by: int,
+    created_date: datetime = None,
+    old: Model = None,
+):
+    if created_date is None:
+        created_date = now()
+
+    if isinstance(objects, (QuerySet, list, tuple)):
+        if old is not None:
+            raise ValueError("diff is not available when objects is a list or queryset")
+        for obj in objects:
+            object_model = obj.__class__.__name__
+            create_consultation_event_task.apply_async(
```

Comment:
Events should always be generated live inside a transaction, No background worker should be involved.

Reply:
We want to be 100% certain that a change has an event associated with it. let's build it that way from the beginning.